test/groovy/util/JenkinsDeleteFileRule.groovy

```diff
@@ -0,0 +1,78 @@
+package util
+
+import com.lesfurets.jenkins.unit.BasePipelineTest
+import org.junit.rules.TestRule
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+import org.yaml.snakeyaml.Yaml
+
+class JenkinsDeleteFileRule implements TestRule {
+
+    final BasePipelineTest testInstance
+    /**
+     * The list of file paths that should have been deleted.
+     * Can be used in tests to assert that productive code
+     * removed the correct files.
+     */
+    List<String> deletedFiles = new ArrayList<>()
+
+    private Boolean skip
+
+    JenkinsDeleteFileRule(BasePipelineTest testInstance) {
+        this.testInstance = testInstance
+    }
+
+    /**
+     * Allows you to skip the actual deletion for test purposes.
+     * @param skip - if `true` will skip the actual deletion.
+     * @return this rule.
+     */
+    JenkinsDeleteFileRule skipDeletion(Boolean skip) {
+        this.skip = skip
+        return this
+    }
+
+    @Override
+    Statement apply(Statement base, Description description) {
+        return statement(base)
+    }
+
+    private Statement statement(final Statement base) {
+        return new Statement() {
+            @Override
+            void evaluate() throws Throwable {
+                testInstance.helper.registerAllowedMethod("deleteFile", [Map], { Map m ->
+                    String path = m?.path
+                    if(!path) {
+                        error "[DeleteFile] File path must not be null or empty."
+                    }
+
+                    deletedFiles.add(path)
+                    deleteFile(path, skip)
+                })
+
+                base.evaluate()
+            }
+        }
+    }
+
+    /**
+     * Deleting the file, skipping deletion if specified.
+     * @param filePath - the path of the file to delete.
+     * @param skipDeletion - if `true` will skip actual deletion.
+     */
+    private void deleteFile(String filePath, Boolean skipDeletion) {
```

Comment:
`deleteFile` takes a single parameter: the path. The test rule keeps track of all the files that have been deleted in a map, so that you can - as you requested in another comment - call `deleteFile` several times and have the checks whether all files have been deleted at the end of the test. 

I changed the behaviour slightly now, and made the skipping of actual deletion the default behaviour. Nonetheless, there is still a method now that allows configuration of the `JenkinsDeleteFileRule` to do **actual** deletion. I found this quite useful, especially, if you are locally trying to verify *how* (not *that*) the solution works, when it runs on Jenkins.  
Especially when there is logic that relies on files existing / not existing, this is helpful, I think.

Reply:
Regarding mocking behaviour: The mock has - let's say - two directions. 

1.) The directions towads the caller. Here we basically return values and exceptions. It is important to be precise with the mock bahaviour. This must reflect the behaviour of the step/plugin/whatever which gets mocked.

2.) The other direction: All the status changes triggered by the call to the mocked entity (e.g. file deleted on FileSystem, perform an http call, whatever ...). There I doubt it makes sense to mimic in an precise way how the mocked part behaves. For testing this we have basically the integration tests. All that is - to my understanding - behind the mock barrier and here we simply _assume_ in the unit tests that an action has been performed. It is in my opinion sufficient to provide the information through the mock that there was an attempt to e.g. delete a file with a certain name so that we can check that later in the assertions of the test.

> logic that relies on files existing / not existing,

Here it is IMO better to mock the call which check that the file exists. That is more straight forwards compared to setting up that status in the file system and configure the tests to use the part of the file system where the status has been setup.

Reply:
In theory, I agree with all of what you said. In practice, however, - and take this as the experience of someone who jumped into Piper and Jenkins Library one and a half weeks ago - you are easily overwhelmed by an abundance of new technologies and concepts. Additionally, it takes quite a while until you have a setup running that allows you to test things locally without wasting too much on development turnaround times. 

To be able to quickly, try out *locally* (from within your IDE) in a test, how something will behave in a real life setup, to see how things really work, can be an enormeous help.  
That said, even the current implementation of `JenkinsReadYamlRule` actually loads a Yaml file making a call to the `Yaml` library. Following your chain of arguments, that rule should not do that, but return a mock Yaml all the time. I assume there are similar reasons to ones I mentioned above, that made the developers choose otherwise.

The way `JenkinsDeleteFileRule` is written now, it allows you do do exactly what you were saying **plus** it actually lets you delete a file, if you choose to configure it so.  I am not saying you should rely on actual file deletion in your final unit tests - but I value the possibility to do so, if I need it.